src/django_components/dependencies.py

```diff
@@ -0,0 +1,764 @@
+#    process all the HTML dependency comments (created in
+#    previous step), obtaining ALL JS and CSS scripts
+#    required by this HTML document. And post-process them,
+#    so the scripts are either inlined into the HTML, or
+#    fetched when the HTML is loaded in the browser.
+#########################################################
+
+
+TContent = TypeVar("TContent", bound=Union[bytes, str])
+
+
+CSS_DEPENDENCY_PLACEHOLDER = '<link name="CSS_PLACEHOLDER">'
+JS_DEPENDENCY_PLACEHOLDER = '<script name="JS_PLACEHOLDER"></script>'
+
+CSS_PLACEHOLDER_BYTES = bytes(CSS_DEPENDENCY_PLACEHOLDER, encoding="utf-8")
+JS_PLACEHOLDER_BYTES = bytes(JS_DEPENDENCY_PLACEHOLDER, encoding="utf-8")
+
+COMPONENT_DEPS_COMMENT = "<!-- _RENDERED {data} -->"
+# E.g. `<!-- _RENDERED table,123 -->`
+COMPONENT_COMMENT_REGEX = re.compile(rb"<!-- _RENDERED (?P<data>[\w\-,/]+?) -->")
+# E.g. `table,123`
+SCRIPT_NAME_REGEX = re.compile(rb"^(?P<comp_cls_hash>[\w\-\./]+?),(?P<id>[\w]+?)$")
+PLACEHOLDER_REGEX = re.compile(
+    r"{css_placeholder}|{js_placeholder}".format(
+        css_placeholder=CSS_DEPENDENCY_PLACEHOLDER,
+        js_placeholder=JS_DEPENDENCY_PLACEHOLDER,
+    ).encode()
+)
+
+
+def render_dependencies(content: TContent, kind: RenderKind = "document") -> TContent:
+    """
+    Given a string that contains parts that were rendered by components,
+    this function inserts all used JS and CSS.
+
+    By default, the string is parsed as an HTML and:
+    - CSS is inserted at the end of `<head>` (if present)
+    - JS is inserted at the end of `<body>` (if present)
+
+    If you used `{% component_dependencies %}`, `{% component_js_dependencies %}`
+    or `{% component_css_dependencies %}`, then the JS and CSS will be inserted
+    only at these locations.
+
+    Example:
+    ```python
+    def my_view(request):
+        template = Template('''
+            {% load components %}
+            <!doctype html>
+            <html>
+                <head></head>
+                <body>
+                    <h1>{{ table_name }}</h1>
+                    {% component "table" name=table_name / %}
+                </body>
+            </html>
+        ''')
+
+        html = template.render(
+            Context({
+                "table_name": request.GET["name"],
+            })
+        )
+
+        # This inserts components' JS and CSS
+        processed_html = render_dependencies(html)
+
+        return HttpResponse(processed_html)
+    ```
+    """
+    is_safestring = isinstance(content, SafeString)
+
+    if isinstance(content, str):
+        content_ = content.encode()
+    else:
+        content_ = cast(bytes, content)
+
+    content_, js_dependencies, css_dependencies = _process_dep_declarations(content_, kind)
+
+    # Replace the placeholders with the actual content
+    did_find_js_placeholder = False
+    did_find_css_placeholder = False
+
+    def on_replace_match(match: "re.Match[bytes]") -> bytes:
```

Comment:
Similar replace logic was previously in the middleware. However, there we replaced only the first occurence of the CSS or JS placeholder.

IMO this is probably the cause this bug https://github.com/EmilStenstrom/django-components/issues/277. So instead, now we will replace ALL placeholders we encounter, so the behaviour is more predictable.